proof/refine/ARM/Finalise_R.thy

```diff
@@ -4362,62 +4489,388 @@ global_interpretation delete_one
    apply auto
   done
 
+(*FIXME RT: move*)
+lemma corres_split_new:
+  assumes x: "corres_underlying sr nf nf' r' P P' a c"
+  assumes y: "\<And>rv rv'. r' rv rv' \<Longrightarrow> corres_underlying sr nf nf' r (R rv) (R' rv') (b rv) (d rv')"
+  assumes    "\<lbrace>Q\<rbrace> a \<lbrace>R\<rbrace>" "\<lbrace>Q'\<rbrace> c \<lbrace>R'\<rbrace>"
+  shows      "corres_underlying sr nf nf' r (P and Q) (P' and Q') (a >>= (\<lambda>rv. b rv)) (c >>= (\<lambda>rv'. d rv'))"
+  using assms
+  apply -
+  apply (rule corres_split')
+     apply (rule corres_guard_imp, rule x, simp_all)
+    apply (erule y)
+   apply (rule hoare_weaken_pre, assumption)
+   apply simp
+  apply (rule hoare_weaken_pre, assumption)
+  apply simp
+  done
+
+lemma set_tcb_obj_ref_corresT:
+  assumes x: "\<And>tcb tcb'. tcb_relation tcb tcb' \<Longrightarrow>
+                         tcb_relation (f (\<lambda>_. new) tcb) (f' tcb')"
+  assumes y: "\<And>tcb. \<forall>(getF, setF) \<in> ran tcb_cap_cases. getF (f (\<lambda>_. new) tcb) = getF tcb"
+  assumes z: "\<forall>tcb. \<forall>(getF, setF) \<in> ran tcb_cte_cases.
+                 getF (f' tcb) = getF tcb"
+  shows      "corres dc (tcb_at t)
+                        (tcb_at' t)
+                    (set_tcb_obj_ref f t new) (threadSet f' t)"
+  by (clarsimp simp: set_tcb_obj_ref_thread_set threadset_corresT x y z)
+
+lemma set_sc_obj_ref_no_stack_update_corres:
```

Comment:
Is this just moved from KHeap_R? I know there's something very similar, maybe just phrased in terms of `update_sched_context` (set_sc_obj_ref` is just an abbreviation of that, I think)

Reply:
Maybe this is new, but just move convenient for you. In which case, maybe it can be moved to live with its cousins

Reply:
yes, there is `update_sc_no_reply_stack_update_ko_at’_corres` in KHeap_R, which I think covers this case. Just having a look at it, we should remove `pspace_aligned` and `pspace_distinct` from the abstract guard because we don’t need to cross `sc_at’` when we already have `ko_at’` for an SC.

Reply:
Thanks for pointing this out. I've removed my version and updated `update_sc_no_reply_stack_update_ko_at’_corres`.